File: appbarspring/src/main/java/com/google/android/material/appbar/TabScrimHelper.java
package com.google.android.material.appbar;


import android.animation.ValueAnimator;
import android.content.res.ColorStateList;
import android.graphics.Color;

import androidx.core.graphics.ColorUtils;
import androidx.core.view.ViewCompat;

import com.google.android.material.animation.AnimationUtils;
import com.google.android.material.tabs.TabLayout;

public class TabScrimHelper implements AppBarLayout.OnOffsetChangedListener {
    private static final int DEFAULT_SCRIM_ANIMATION_DURATION = 600;

    private TabLayout mTabLayout;
    private CollapsingToolbarLayout mToolbarLayout;
    private long mScrimAnimationDuration;
    private int mScrimAlpha;
    private boolean mScrimsAreShown;
    private ValueAnimator mScrimAnimator;
    private final int mNormalColor;
    private final int mSelectedColor;
    private int mCollapseTabSelectTextColor;
    private int mCollapseTabNormalTextColor;
    private int mCollapseTabBackgroundColor;

    public TabScrimHelper(TabLayout tabLayout, CollapsingToolbarLayout toolbarLayout) {
        initDefault();
        mTabLayout = tabLayout;
        mToolbarLayout = toolbarLayout;
        ColorStateList colorStateList = mTabLayout.getTabTextColors();
        mSelectedColor = colorStateList.getColorForState(new int[]{android.R.attr.state_selected}, Color.parseColor("#3F51B5"));
        mNormalColor = colorStateList.getDefaultColor();
    }

    @SuppressWarnings("unused")
    public void setCollapseTabBackgroundColor(int collapseTabBackgroundColor) {
        mCollapseTabBackgroundColor = collapseTabBackgroundColor;
    }

    @SuppressWarnings("unused")
    public void setCollapseTabSelectTextColor(int collapseTabSelectTextColor) {
        mCollapseTabSelectTextColor = collapseTabSelectTextColor;
    }

    private void initDefault() {
        mScrimAnimationDuration = DEFAULT_SCRIM_ANIMATION_DURATION;
        mCollapseTabBackgroundColor = Color.parseColor("#3F51B5");
        mCollapseTabNormalTextColor = Color.parseColor("#FFFFFF");
        mCollapseTabSelectTextColor = Color.parseColor("#FF4081");
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        setScrimsShown(mToolbarLayout.getHeight() + verticalOffset < mToolbarLayout.getScrimVisibleHeightTrigger());

    }

    public void setScrimsShown(boolean shown) {
        setScrimsShown(shown, ViewCompat.isLaidOut(mToolbarLayout) && !mToolbarLayout.isInEditMode());
    }

    public void setScrimsShown(boolean shown, boolean animate) {
        if (mScrimsAreShown != shown) {
            if (animate) {
                animateScrim(shown ? 0xFF : 0x0);
            } else {
                setScrimAlpha(shown ? 0xFF : 0x0);
            }
            mScrimsAreShown = shown;
        }
    }

    private void animateScrim(int targetAlpha) {
        if (mScrimAnimator == null) {
            mScrimAnimator = new ValueAnimator();
            mScrimAnimator.setDuration(mScrimAnimationDuration);
            mScrimAnimator.setInterpolator(
                    targetAlpha > mScrimAlpha
                            ? AnimationUtils.FAST_OUT_LINEAR_IN_INTERPOLATOR
                            : AnimationUtils.LINEAR_OUT_SLOW_IN_INTERPOLATOR);
            mScrimAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator animator) {
                    setScrimAlpha((Integer) animator.getAnimatedValue());
                }
            });
        } else if (mScrimAnimator.isRunning()) {
            mScrimAnimator.cancel();
        }

        mScrimAnimator.setIntValues(mScrimAlpha, targetAlpha);
        mScrimAnimator.start();
    }

    void setScrimAlpha(int alpha) {
        if (alpha != mScrimAlpha) {
            mScrimAlpha = alpha;
            updateLayout();
        }
    }

    private void updateLayout() {
        int color = ColorUtils.setAlphaComponent(mCollapseTabBackgroundColor, mScrimAlpha);
        mTabLayout.setBackgroundColor(color);

        float i = 1.f * mScrimAlpha / 255;
        mTabLayout.setTabTextColors(
                ColorUtils.blendARGB(mNormalColor, mCollapseTabNormalTextColor, i),
                ColorUtils.blendARGB(mSelectedColor, mCollapseTabSelectTextColor, i));

    }
}

File: appbarspring/src/main/java/com/google/android/TabScrimHelper.kt
package com.google.android


import android.animation.ValueAnimator
import android.graphics.Color
import androidx.core.graphics.ColorUtils
import androidx.core.view.ViewCompat
import com.google.android.material.animation.AnimationUtils
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.tabs.TabLayout

class TabScrimHelper(private val mTabLayout: TabLayout, private val mToolbarLayout: CollapsingToolbarLayout) : AppBarLayout.OnOffsetChangedListener {


    private var mScrimAnimationDuration: Long = 0
    private var mScrimAlpha: Int = 0
    private var mScrimsAreShown: Boolean = false
    private var mScrimAnimator: ValueAnimator? = null
    private val mNormalColor: Int
    private val mSelectedColor: Int
    private var mCollapseTabSelectTextColor: Int = 0
    private var mCollapseTabNormalTextColor: Int = 0
    private var mCollapseTabBackgroundColor: Int = 0

    init {
        initDefault()
        val colorStateList = mTabLayout.tabTextColors
        mSelectedColor = colorStateList!!.getColorForState(intArrayOf(android.R.attr.state_selected), Color.parseColor("#3F51B5"))
        mNormalColor = colorStateList.defaultColor
    }


    private fun initDefault() {
        mScrimAnimationDuration = DEFAULT_SCRIM_ANIMATION_DURATION.toLong()
        mCollapseTabBackgroundColor = Color.parseColor("#3F51B5")
        mCollapseTabNormalTextColor = Color.parseColor("#FFFFFF")
        mCollapseTabSelectTextColor = Color.parseColor("#FF4081")
    }

    override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
        setScrimsShown(mToolbarLayout.height + verticalOffset < mToolbarLayout.scrimVisibleHeightTrigger)

    }

    private fun setScrimsShown(shown: Boolean, animate: Boolean = ViewCompat.isLaidOut(mToolbarLayout) && !mToolbarLayout.isInEditMode) {
        if (mScrimsAreShown != shown) {
            if (animate) {
                animateScrim(if (shown) 0xFF else 0x0)
            } else {
                setScrimAlpha(if (shown) 0xFF else 0x0)
            }
            mScrimsAreShown = shown
        }
    }

    private fun animateScrim(targetAlpha: Int) {
        if (mScrimAnimator == null) {
            mScrimAnimator = ValueAnimator()
            mScrimAnimator!!.duration = mScrimAnimationDuration
            mScrimAnimator!!.interpolator = if (targetAlpha > mScrimAlpha)
                AnimationUtils.FAST_OUT_LINEAR_IN_INTERPOLATOR
            else
                AnimationUtils.LINEAR_OUT_SLOW_IN_INTERPOLATOR
            mScrimAnimator!!.addUpdateListener { animator -> setScrimAlpha(animator.animatedValue as Int) }
        } else if (mScrimAnimator!!.isRunning) {
            mScrimAnimator!!.cancel()
        }

        mScrimAnimator!!.setIntValues(mScrimAlpha, targetAlpha)
        mScrimAnimator!!.start()
    }


    private fun updateLayout() {
        val color = ColorUtils.setAlphaComponent(mCollapseTabBackgroundColor, mScrimAlpha)
        mTabLayout.setBackgroundColor(color)

        val i = 1f * mScrimAlpha / 255
        mTabLayout.setTabTextColors(
                ColorUtils.blendARGB(mNormalColor, mCollapseTabNormalTextColor, i),
                ColorUtils.blendARGB(mSelectedColor, mCollapseTabSelectTextColor, i))

    }

    companion object {
        private const val DEFAULT_SCRIM_ANIMATION_DURATION = 600
    }


    fun setCollapseTabBackgroundColor(collapseTabBackgroundColor: Int) {
        mCollapseTabBackgroundColor = collapseTabBackgroundColor
    }

    fun setCollapseTabSelectTextColor(collapseTabSelectTextColor: Int) {
        mCollapseTabSelectTextColor = collapseTabSelectTextColor
    }

    fun setScrimAlpha(alpha: Int) {
        if (alpha != mScrimAlpha) {
            mScrimAlpha = alpha
            updateLayout()
        }
    }

}
